### PR TITLE
Set compileSdk and targetSdk to 33 (Android 13)

### PR DESCRIPTION
**Pull request by Stypox**
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
`android:exported` in now required in the manifest on all activities/services/receivers/providers. It was set to `true` for those that need to interact with outside apps or the OS, while others have `exported=false`. Changing the target sdk also required updating LeakCanary to the latest version as the older version being used was not using `android:exported` in `AndroidManifest.xml`.

I don't know whether problems could arise just because of this number increase, maybe @TacoTheDank can shed some light on it?

#### Fixes the following issue(s)

- https://github.com/TeamNewPipe/NewPipe/issues/9292#issuecomment-1304571423

#### Relies on the following changes
- #9333

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tsiflimagas** [comment]
Opening any video crashes the app.

## Exception
* __User Action:__ ui error
* __Request:__ ACRA report
* __Content Country:__ US
* __Content Language:__ en-US
* __App Language:__ en_US
* __Service:__ none
* __Version:__ 0.24.0
* __OS:__ Linux Android 13 - 33
<details><summary><b>Crash log </b></summary><p>

```
java.lang.IllegalArgumentException: org.schabi.newpipe.debug.targetapi33: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
	at android.app.PendingIntent.checkFlags(PendingIntent.java:401)
	at android.app.PendingIntent.getBroadcastAsUser(PendingIntent.java:671)
	at android.app.PendingIntent.getBroadcast(PendingIntent.java:658)
	at org.schabi.newpipe.player.notification.NotificationUtil.createNotification(NotificationUtil.java:136)
	at org.schabi.newpipe.player.notification.NotificationUtil.createNotificationAndStartForeground(NotificationUtil.java:185)
	at org.schabi.newpipe.player.notification.NotificationPlayerUi.initPlayer(NotificationPlayerUi.java:32)
	at org.schabi.newpipe.player.Player$$ExternalSyntheticLambda0.accept(Unknown Source:2)
	at j$.util.Iterator$-CC.$default$forEachRemaining(Iterator.java:116)
	at j$.util.Iterator$-EL.forEachRemaining(Unknown Source:10)
	at j$.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at j$.util.stream.ReferencePipeline$Head.forEachOrdered(ReferencePipeline.java:590)
	at org.schabi.newpipe.player.ui.PlayerUiList.call(PlayerUiList.java:88)
	at org.schabi.newpipe.player.Player.initPlayer(Player.java:522)
	at org.schabi.newpipe.player.Player.initPlayback(Player.java:487)
	at org.schabi.newpipe.player.Player.lambda$handleIntent$2$org-schabi-newpipe-player-Player(Player.java:425)
	at org.schabi.newpipe.player.Player$$ExternalSyntheticLambda27.run(Unknown Source:16)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeCallbackObserver.onComplete(MaybeCallbackObserver.java:93)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeObserveOn$ObserveOnMaybeObserver.run(MaybeObserveOn.java:105)
	at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7898)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)

```
</details>
<hr>

**Comment by Nizami20052022** [comment]
If this pull request was merged, a new version would be released.

**Comment by TobiGr** [comment]
@Nizami20052022 Please be are that you are creating a lot of comments and new issues. Many of them do not any value to the project or discussions and are just consuming resources for reviewing and replying. Our resources are limited. Please make sure to stick to our [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md). If you fail to do so, we might need to restrict your ability to interact with our repositories to not bind any more resources.

**Comment by Nizami20052022** [comment]
I haven't opened many issues.😪

**Comment by opusforlife2** [comment]
@Nizami20052022 It's more about the amount of effort it takes to deal with it. If a lot of your opened issues are being closed, and your comments being marked off-topic, that's an indication for you to read more and comment less. It will improve the quality of your interactions in FOSS repos in general, too.

**Comment by Nizami20052022** [comment]
If you allow me, I can create my own pull request.

**Comment by opusforlife2** [comment]
PRs are always welcome, as we've accepted yours previously. There's no permission needed. Just make sure the code is of good quality and follow our contribution guidelines.

**Comment by Nizami20052022** [comment]
I can contribute other than translation.I finished the translation.

**Comment by AudricV** [comment]
> I don't know whether problems could arise just because of this number increase, maybe TacoTheDank can shed some light on it?

I recommend reading Android version differences ([API 30](https://developer.android.com/about/versions/11/behavior-changes-all) - [changes when targeting this version](https://developer.android.com/about/versions/11/behavior-changes-11), [API 31](https://developer.android.com/about/versions/12/behavior-changes-all) - [changes when targeting this version](https://developer.android.com/about/versions/12/behavior-changes-12), [API 32](https://developer.android.com/about/versions/12/12L/summary#what-to-test) and [API 33](https://developer.android.com/about/versions/13/behavior-changes-all) - [changes when targeting this version](https://developer.android.com/about/versions/13/behavior-changes-13)) and changelog of Jetpack libraries updates, to give us an idea about what changes are required.

**Comment by TacoTheDank** [comment]
I second @AudricV's suggestion.

**Comment by Isira-Seneviratne** [comment]
> 



> Opening any video crashes the app.
> 
> ## Exception
> * **User Action:** ui error
> * **Request:** ACRA report
> * **Content Country:** US
> * **Content Language:** en-US
> * **App Language:** en_US
> * **Service:** none
> * **Version:** 0.24.0
> * **OS:** Linux Android 13 - 33
> 
> Crash log

The `PendingIntent` mutability flags are required on Android 12 and higher, and the `PendingIntent`s created don't have the flags set.

**Comment by antmonteiro** [comment]
You may also want to update `com.android.tools:desugar_jdk_libs` to 1.1.8 to target  API level 33. (https://github.com/google/desugar_jdk_libs/blob/master/CHANGELOG.md)

**Comment by Isira-Seneviratne** [comment]
> You may also want to update `com.android.tools:desugar_jdk_libs` to 1.1.8 to target  API level 33. (https://github.com/google/desugar_jdk_libs/blob/master/CHANGELOG.md)

That's done here: #9285

**Comment by Stypox** [comment]
I pushed two commits making sure that NewPipe now requests the notification permission, as explained [here](https://developer.android.com/about/versions/13/behavior-changes-13#notification-permission-fgs-appearance).

On Android 13+ we need to specify player notification actions with the `PlaybackState` class, see [here](https://developer.android.com/about/versions/13/behavior-changes-13#playback-controls). This can be done in a separate PR I think, since currently the custom actions are not shown on Android 13 regardless of whether the app targets 29 or 33.

**Comment by Stypox** [comment]
The code smells are unrelated

**Comment by AudricV** [comment]
> The code smells are unrelated

No, they are: see https://developer.android.com/reference/android/app/Activity.html?hl=en#onBackPressed() and https://developer.android.com/reference/android/content/Intent?hl=en#getSerializableExtra(java.lang.String)

These methods are deprecated since API 33 and used in `MainActivity`.

**Comment by Stypox** [comment]
I don't think I am going to fix those deprecations here.

For example, `getSerializableExtra(String)` suggests `getSerializableExtra(String, class)` as a replacement, but the latter requires API 33 and I couldn't find a similar method in `IntentCompat`, so it's not solvable as of now. If you or @Isira-Seneviratne find a solution feel free to push commits to this PR and then merge it; otherwise please merge it as it is and we will fix deprecations once `IntentCompat` catches up.

image

As for the `onBackPressed` deprecations, I am not sure how to proceed. There seems to be too many changes to do in this PR. I think it's better to do them in another PR.

**Comment by Stypox** [comment]
I took a deeper look into the `onBackPressed` deprecation. When we just need to always consume the `onBack` it is simple to switch to just register a callback that is always called. But when you have a conditional `onBack` (e.g. close the drawer or the search bar if it is open, otherwise let `super` handle the `onBack`), this [requires](https://stackoverflow.com/a/73804521/9481500) enabling or disabling the callback based on whether the drawer or the search bar are open. This means having to continuously sync the "openness" state of drawer/search with the "enabledness" state of the `onBack` callback. Not only does this require much more code than before, it also greatly enlarges the possibility for programming errors and non-synced state. Using enableable/disableable `onBack` callbacks in a state-based UI system like Jetpack Compose would work perfectly well, but for us it would be a nightmare. So I will merge this PR as it is.

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -85,20 +93,26 @@
...
             android:name=".download.DownloadActivity"
+            android:exported="false"
             android:label="@string/app_name"
             android:launchMode="singleTask" />
 
         <service android:name="us.shandian.giga.service.DownloadManagerService" />
 
         <activity
             android:name=".util.FilePickerActivityHelper"
+            android:exported="true"
```
Comment (ghost) [inline app/src/main/AndroidManifest.xml:115]: @Stypox why this one is set to true? I understand why other have their specific values but not this one. It seems to me it should be false?

**Pull request merged**